src/routes/updateProfile.ts

```diff
@@ -25,174 +25,182 @@ export const updateProfile = async (
+    })
   }
 
   // Get existing profile.
-  let existingProfile: Profile = { ...EMPTY_PROFILE };
+  let existingProfile: Profile = { ...EMPTY_PROFILE }
+  let existingProfileId: number | undefined
   try {
-    const stringifiedData = await env.PROFILES.get(getProfileKey(publicKey));
-    if (stringifiedData) {
-      existingProfile = JSON.parse(stringifiedData);
+    const _profile = await getProfileFromPublicKey(env, publicKey)
+    if (_profile) {
+      existingProfile = _profile
+      existingProfileId = _profile.id
     }
   } catch (err) {
-    console.error("Profile retrieval or parsing", err);
+    console.error('Profile retrieval', err)
 
     return respond(500, {
-      error: "Failed to retrieve or parse existing profile",
-      message: err instanceof Error ? err.message : `${err}`,
-    });
+      error:
+        'Failed to retrieve existing profile: ' +
+        (err instanceof Error ? err.message : `${err}`),
+    })
   }
 
   // Validate nonce to prevent replay attacks.
   if (requestBody.profile.nonce !== existingProfile.nonce) {
     return respond(401, {
-      error: "Invalid body",
-      message: `Invalid nonce. Expected: ${existingProfile.nonce}`,
-    });
+      error: `Invalid nonce. Expected: ${existingProfile.nonce}`,
+    })
   }
 
-  const { name, nft } = requestBody.profile;
+  const { name, nft } = requestBody.profile
 
   // If setting name, verify unique.
-  if (typeof name === "string") {
+  if (typeof name === 'string') {
     try {
-      if (await env.PROFILES.get(getPublicKeyForNameTakenKey(name))) {
+      const namedProfile = await getProfileFromName(env, name)
+      // Only error if profile that is not the existing profile has the name.
+      if (namedProfile && namedProfile.id !== existingProfileId) {
         return respond(500, {
-          error: "Invalid name",
-          message: "Name already exists.",
-        });
+          error: 'Name already taken.',
+        })
       }
     } catch (err) {
-      console.error("Name uniqueness retrieval", err);
+      console.error('Name uniqueness retrieval', err)
 
       return respond(500, {
-        error: "Failed to check name uniqueness",
-        message: err instanceof Error ? err.message : `${err}`,
-      });
+        error:
+          'Failed to check name uniqueness: ' +
+          (err instanceof Error ? err.message : `${err}`),
+      })
     }
   }
 
-  // If setting NFT, verify it belongs to the public key.
+  // If setting NFT, verify it belongs to the profile's public key.
   if (nft) {
     try {
+      // If profile exists, get public key for the NFT's chain.
+      const chainPublicKey = existingProfileId
+        ? (
+            await getPreferredProfilePublicKey(
+              env,
+              existingProfileId,
+              nft.chainId
+            )
+          )?.publicKey
+        : // If profile doesn't exist yet but the NFT is on a chain being added right now, use the current public key.
```

Comment:
Again, when might this happen? You don't have a profile, you're creating one, adding an NFT as a profile picture...?

Reply:
the very first time this route is called, when no profile exists for the public key: `getPreferredProfilePublicKey` would fail because no profile exists, but the current public key creating a profile is valid since it's the one creating the profile and will be the first public key registered

Reply:
Got it. A comment to this effect could be helpful for readers.